iothub_client/samples/iothub_ll_c2d_sample/iothub_ll_c2d_sample.c

```diff
@@ -43,13 +43,55 @@
 #include "certs.h"
 #endif // SET_TRUSTED_CERT_IN_SAMPLES
 
+// Uncomment this define to use Asynchronous ACK of Cloud-to-Device messages.
+// #define USE_C2D_ASYNC_ACK
+
+#ifdef USE_C2D_ASYNC_ACK
+#include "azure_c_shared_utility/singlylinkedlist.h"
+#endif
+
 /* Paste in the your iothub connection string  */
 static const char* connectionString = "[device connection string]";
 
 #define MESSAGE_COUNT        3
 static bool g_continueRunning = true;
 static size_t g_message_recv_count = 0;
 
+#ifdef USE_C2D_ASYNC_ACK
+static SINGLYLINKEDLIST_HANDLE g_cloudMessages;
+
+static bool ack_and_remove_message(const void* item, const void* match_context, bool* continue_processing)
+{
+    IOTHUB_DEVICE_CLIENT_LL_HANDLE device_ll_handle = (IOTHUB_DEVICE_CLIENT_LL_HANDLE)match_context;
+    IOTHUB_MESSAGE_HANDLE message = (IOTHUB_MESSAGE_HANDLE)item;
+
+    const char* messageId;
+    if ((messageId = IoTHubMessage_GetMessageId(message)) == NULL)
+    {
+        messageId = "<unavailable>";
+    }
+
+    (void)printf("Sending ACK for cloud message (Message ID: %s)\r\n", messageId);
+
+    if (IoTHubDeviceClient_LL_SendMessageDisposition(device_ll_handle, message, IOTHUBMESSAGE_ACCEPTED) != IOTHUB_CLIENT_OK)
+    {
+        (void)printf("ERROR: IoTHubDeviceClient_LL_SendMessageDisposition..........FAILED!\r\n");
+        // If (and only if) IoTHubDeviceClient_LL_SendMessageDisposition fails, the IOTHUB_MESSAGE_HANDLE must be explicitly destroyed.
+        IoTHubMessage_Destroy(message);
```

Comment:
Having either the SDK or the app manage memory is confusing. 
In this case it's probably better to assert / crash instead of deallocating something that will definitely block the protocol maybe causing Hub to disconnect. 

Going out of sync on MQTT or AMQP is a dangerous situation.

Reply:
Changes made to sample and documentation.